File: src/simcore/state.py
"""Immutable state primitives: ``Stock`` and the per-step ``State`` snapshot.

Both are frozen (decision: the engine writes a *new* ``State`` each step via
``dataclasses.replace`` rather than mutating in place — see the reconciled Frozen
API note in the plan). ``State.stocks`` is wrapped in a ``MappingProxyType`` so a
snapshot cannot be mutated through its mapping after construction.

Caveat (per advisor): the proxy guards *mutation*, not *order*. Canonical order
for reductions comes from explicitly sorting by stock/flow id at each reduction
(later steps), never from mapping iteration order.

Pure stdlib only.
"""

import math
from collections.abc import Mapping
from dataclasses import dataclass, field
from types import MappingProxyType

from simcore.ids import DomainId, StockId, UnitLabel
from simcore.quantities import Quantity, StockKind

# Shared immutable empty default for ``State.aux`` (the non-conserved channel,
# Phase-1 P2). A ``MappingProxyType`` is not in dataclass's mutable-default
# blacklist (list/dict/set) and is effectively immutable, so one shared instance
# is a safe default — every ``State`` re-wraps its own copy in ``__post_init__``.
_EMPTY_AUX: Mapping[str, float] = MappingProxyType({})

# Shared "not supplied" sentinel for ``Stock.composition`` (P2.1). Empty means
# "default to the 1:1 ``{quantity: 1.0}`` map" — filled in ``__post_init__`` once
# the stock's own quantity is known. Same safe-shared-MappingProxyType idiom as
# ``_EMPTY_AUX`` (not in dataclass's mutable-default blacklist).
_EMPTY_COMPOSITION: Mapping[Quantity, float] = MappingProxyType({})


@dataclass(frozen=True)
class Stock:
    """A single well-mixed compartment holding an amount of one quantity.

    ``amount`` is in the quantity's canonical unit (the ``unit`` label records
    which; see ``simcore.quantities.CANONICAL_UNIT``). Frozen: updates produce a
    new ``Stock`` via ``dataclasses.replace``.
    """

    id: StockId
    domain: DomainId
    quantity: Quantity
    unit: UnitLabel
    amount: float
    kind: StockKind
    # POPULATION only: at/below this level the stock snaps to 0 and the residual
    # is routed to the numerical-loss sink (decision #6). Ignored for other kinds.
    extinction_threshold: float = 0.0
    # BOUNDARY source (e.g. solar): never throttled by arbitration min-scaling
    # (decision #13). Meaningful only for BOUNDARY stocks.
    unclamped: bool = False
    # Element composition (P2.1): moles of each conserved quantity that one
    # canonical unit of this stock contributes to the conservation ledger. The
    # empty-sentinel default fills to the 1:1 ``{quantity: 1.0}`` map in
    # ``__post_init__``, so every Phase-1 stock is unchanged. A gas-phase stock
    # carries several: CO2={CARBON:1, OXYGEN:2}, O2={OXYGEN:2}; biomass stays
    # {CARBON:1}. ``hash=False`` because a ``Mapping`` is unhashable — but it
    # stays in ``__eq__`` so a dropped/garbled composition is caught by the
    # snapshot round-trip. The conservation gate (``flow``/``conservation``)
    # folds this map; nothing else (arbitration, integrator, observation) reads it.
    composition: Mapping[Quantity, float] = field(
        default=_EMPTY_COMPOSITION, hash=False
    )

    def __post_init__(self) -> None:
        # NaN/Inf are forbidden anywhere in state (determinism/serialization).
        if not math.isfinite(self.amount):
            raise ValueError(f"Stock {self.id!r} amount is not finite: {self.amount!r}")
        if not math.isfinite(self.extinction_threshold):
            raise ValueError(
                f"Stock {self.id!r} extinction_threshold is not finite: "
                f"{self.extinction_threshold!r}"
            )
        # `unclamped` is meaningful only for BOUNDARY sources (decision #13): it
        # tells arbitration's min-scaling never to throttle this stock. An
        # unclamped POOL/POPULATION would silently escape throttling and could go
        # negative — a conservation break — so reject it at construction. (Guard
        # added in step 4 alongside the Boundary domain module; it tightens this
        # step-2 primitive to match its own documented contract.)
        if self.unclamped and self.kind is not StockKind.BOUNDARY:
            raise ValueError(
                f"Stock {self.id!r} is unclamped but kind={self.kind.name}; "
                "unclamped is only valid on BOUNDARY stocks (decision #13)"
            )
        # Element composition (P2.1). Empty (the not-supplied sentinel) → the 1:1
        # ``{quantity: 1.0}`` default. Validate: every key a ``Quantity``, every
        # coeff finite, and the stock's own quantity present with a positive coeff
        # (so a stock always contributes to its nominal quantity). Detach from the
        # caller's mapping and freeze read-only — same discipline as ``stocks``/``aux``.
        composition = dict(self.composition) or {self.quantity: 1.0}
        for q, coeff in composition.items():
            if not isinstance(q, Quantity):
                raise ValueError(
                    f"Stock {self.id!r} composition key {q!r} is not a Quantity"
                )
            if not math.isfinite(coeff):
                raise ValueError(
                    f"Stock {self.id!r} composition[{q.name}] is not finite: {coeff!r}"
                )
        self_coeff = composition.get(self.quantity)
        if self_coeff is None or self_coeff <= 0.0:
            raise ValueError(
                f"Stock {self.id!r} composition must include its own quantity "
                f"{self.quantity.name} with a positive coeff (got {self_coeff!r})"
            )
        # A POPULATION stock must be single-quantity (P2.1 invariant): extinction
        # routes its residual via ``loss_sink_id(stock.quantity)`` — one nominal
        # quantity — so a multi-quantity POPULATION would leak the non-nominal
        # quantities' mass at extinction. Reject at construction.
        if self.kind is StockKind.POPULATION and set(composition) != {self.quantity}:
            raise ValueError(
                f"Stock {self.id!r} is POPULATION but multi-quantity "
                f"(composition keys {sorted(k.name for k in composition)}); a "
                "POPULATION stock must be single-quantity (extinction loss-sink "
                "routing is single-quantity, P2.1)"
            )
        object.__setattr__(self, "composition", MappingProxyType(composition))


@dataclass(frozen=True)
class State:
    """One immutable simulation snapshot.

    Time is the integer step count ``n`` (decision #14): wall time is ``n * dt``,
    evaluated, never accumulated. ``rng_seed`` is carried in state so draws are
    keyed by ``(seed, key, n)`` and stay order-independent (decision #12).

    Note: a ``State`` is **not hashable** — the ``MappingProxyType`` stocks field
    is unhashable, so ``hash(state)`` raises. Equality (by contents) works. If a
    later test needs states in a ``set``/dict-key, give ``State`` a custom
    ``__hash__`` then (e.g. over ``n`` + sorted stock ids); don't rely on it now.

    ``aux`` is the **non-conserved auxiliary channel** (Phase-1 P2): scalar
    accumulators (thermal time, …) advanced by the integrator in parallel to
    ``stocks`` but **outside** the conservation gate — they have no balanced
    counterparty, so they are not flows and are invisible to
    ``conservation.compute_ledger`` (which reasons only over ``stocks``). The
    empty default keeps every pre-P2 call site (and the Phase-0/0.5 goldens, modulo
    the schema bump) unchanged. Keys are stable, canonical-sortable names; values
    are advanced by ``simcore.auxiliary`` processes (see the integrator).
    """

    n: int
    stocks: Mapping[StockId, Stock]
    rng_seed: int
    aux: Mapping[str, float] = _EMPTY_AUX

    def __post_init__(self) -> None:
        if self.n < 0:
            raise ValueError(f"State.n must be >= 0, got {self.n}")
        # Each mapping key must equal its stock's own id — a cheap guard against
        # a stock filed under the wrong key (which would corrupt id-keyed lookups
        # and canonical-order reductions).
        for key, stock in self.stocks.items():
            if key != stock.id:
                raise ValueError(f"State.stocks key {key!r} != stock.id {stock.id!r}")
        # NaN/Inf are forbidden anywhere in state (determinism/serialization) — the
        # same isfinite-only discipline as ``Stock.amount``; aux values are plain
        # accumulators, so no further coercion.
        for name, value in self.aux.items():
            if not math.isfinite(value):
                raise ValueError(f"State.aux[{name!r}] is not finite: {value!r}")
        # Detach from the caller's dicts and forbid post-construction mutation.
        object.__setattr__(self, "stocks", MappingProxyType(dict(self.stocks)))
        object.__setattr__(self, "aux", MappingProxyType(dict(self.aux)))

File: rust/crates/simcore/src/state.rs
//! Immutable state primitives: [`Stock`] and the per-step [`State`] snapshot — the
//! Rust port of `simcore.state`.
//!
//! Python freezes both dataclasses and re-fires every invariant in `__post_init__`,
//! including on `dataclasses.replace`. The Rust port mirrors that: construction goes
//! through [`Stock::new`] / [`State::new`], which validate and return `Result`, and
//! the amounts-only "replace" helpers ([`Stock::with_amount`], the `State` rebuilds
//! in the integrator) re-run [`State::new`] so the same guards re-fire on every step.
//!
//! Composition (P2.1): the empty map means "default to the 1:1 `{quantity: 1.0}`
//! map", filled in [`Stock::new`] once the stock's own quantity is known — so a stock
//! always contributes to its nominal quantity. A gas-phase stock carries several
//! (CO₂ = `{Carbon:1, Oxygen:2}`). Composition iteration order never affects a float
//! result (each quantity is an independent accumulator downstream), so the
//! `BTreeMap` container's ordering is a convenience, not a correctness dependency.

use std::collections::BTreeMap;

use crate::error::SimError;
use crate::ids::{DomainId, StockId, UnitLabel};
use crate::quantities::{Quantity, StockKind};

/// A single well-mixed compartment holding an amount of one quantity.
///
/// Frozen in spirit: updates produce a *new* `Stock` (see [`Stock::with_amount`]).
/// `composition` is always the filled map (never empty) once constructed.
#[derive(Debug, Clone, PartialEq)]
pub struct Stock {
    pub id: StockId,
    pub domain: DomainId,
    pub quantity: Quantity,
    pub unit: UnitLabel,
    pub amount: f64,
    pub kind: StockKind,
    /// POPULATION only: at/below this level the stock snaps to 0 and the residual is
    /// routed to the numerical-loss sink. Ignored for other kinds.
    pub extinction_threshold: f64,
    /// BOUNDARY source only: never throttled by arbitration min-scaling.
    pub unclamped: bool,
    /// Element composition: moles of each conserved quantity one canonical unit of
    /// this stock contributes to the ledger. Always the filled map post-construction.
    pub composition: BTreeMap<Quantity, f64>,
}

impl Stock {
    /// Construct and validate a stock (mirrors Python `Stock.__post_init__`).
    ///
    /// An empty `composition` defaults to the 1:1 `{quantity: 1.0}` map. Returns
    /// [`SimError::Validation`] on any invariant breach (non-finite amount/threshold,
    /// an unclamped non-BOUNDARY stock, a bad composition, a multi-quantity
    /// POPULATION).
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        id: StockId,
        domain: DomainId,
        quantity: Quantity,
        unit: UnitLabel,
        amount: f64,
        kind: StockKind,
        extinction_threshold: f64,
        unclamped: bool,
        composition: BTreeMap<Quantity, f64>,
    ) -> Result<Stock, SimError> {
        if !amount.is_finite() {
            return Err(SimError::Validation(format!(
                "Stock {id:?} amount is not finite: {amount:?}"
            )));
        }
        if !extinction_threshold.is_finite() {
            return Err(SimError::Validation(format!(
                "Stock {id:?} extinction_threshold is not finite: {extinction_threshold:?}"
            )));
        }
        // `unclamped` is meaningful only for BOUNDARY sources: an unclamped
        // POOL/POPULATION would silently escape throttling and could go negative.
        if unclamped && kind != StockKind::Boundary {
            return Err(SimError::Validation(format!(
                "Stock {id:?} is unclamped but kind={}; unclamped is only valid on \
                 BOUNDARY stocks (decision #13)",
                kind.name()
            )));
        }
        // Empty (the not-supplied sentinel) → the 1:1 default.
        let composition = if composition.is_empty() {
            BTreeMap::from([(quantity, 1.0)])
        } else {
            composition
        };
        for (q, coeff) in &composition {
            if !coeff.is_finite() {
                return Err(SimError::Validation(format!(
                    "Stock {id:?} composition[{}] is not finite: {coeff:?}",
                    q.name()
                )));
            }
        }
        match composition.get(&quantity) {
            Some(c) if *c > 0.0 => {}
            other => {
                return Err(SimError::Validation(format!(
                    "Stock {id:?} composition must include its own quantity {} with a \
                     positive coeff (got {other:?})",
                    quantity.name()
                )));
            }
        }
        // A POPULATION stock must be single-quantity: extinction routes its residual
        // via one nominal quantity, so a multi-quantity POPULATION would leak mass.
        if kind == StockKind::Population
            && (composition.len() != 1 || !composition.contains_key(&quantity))
        {
            let mut keys: Vec<&str> = composition.keys().map(|q| q.name()).collect();
            keys.sort_unstable();
            return Err(SimError::Validation(format!(
                "Stock {id:?} is POPULATION but multi-quantity (composition keys \
                 {keys:?}); a POPULATION stock must be single-quantity (extinction \
                 loss-sink routing is single-quantity, P2.1)"
            )));
        }
        Ok(Stock {
            id,
            domain,
            quantity,
            unit,
            amount,
            kind,
            extinction_threshold,
            unclamped,
            composition,
        })
    }

    /// A copy with `amount` replaced (the amounts-only `dataclasses.replace`).
    ///
    /// Re-validates finiteness of the new amount — the only invariant a
    /// change-of-amount can break (kind/composition/unclamped are unchanged and were
    /// valid at construction, so re-checking them would always pass). This matches
    /// the *effect* of Python re-running `__post_init__` on `replace`.
    pub fn with_amount(&self, amount: f64) -> Result<Stock, SimError> {
        if !amount.is_finite() {
            return Err(SimError::Validation(format!(
                "Stock {:?} amount is not finite: {amount:?}",
                self.id
            )));
        }
        Ok(Stock {
            amount,
            ..self.clone()
        })
    }
}

/// One immutable simulation snapshot.
///
/// Time is the integer step count `n` (`t = n*dt`, evaluated, never accumulated).
/// `aux` is the non-conserved auxiliary channel — scalar accumulators advanced by the
/// integrator outside the conservation gate.
#[derive(Debug, Clone, PartialEq)]
pub struct State {
    pub n: u64,
    pub stocks: BTreeMap<StockId, Stock>,
    pub rng_seed: u64,
    pub aux: BTreeMap<String, f64>,
}

impl State {
    /// Construct and validate a snapshot (mirrors Python `State.__post_init__`).
    ///
    /// `n < 0` is unrepresentable (`u64`), subsuming Python's `n >= 0` guard. Each
    /// stock's map key must equal its own id, and every aux value must be finite.
    pub fn new(
        n: u64,
        stocks: BTreeMap<StockId, Stock>,
        rng_seed: u64,
        aux: BTreeMap<String, f64>,
    ) -> Result<State, SimError> {
        for (key, stock) in &stocks {
            if *key != stock.id {
                return Err(SimError::Validation(format!(
                    "State.stocks key {key:?} != stock.id {:?}",
                    stock.id
                )));
            }
        }
        for (name, value) in &aux {
            if !value.is_finite() {
                return Err(SimError::Validation(format!(
                    "State.aux[{name:?}] is not finite: {value:?}"
                )));
            }
        }
        Ok(State {
            n,
            stocks,
            rng_seed,
            aux,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn pool(id: &str, amount: f64) -> Stock {
        Stock::new(
            id.to_string(),
            "d".to_string(),
            Quantity::Carbon,
            "mol".to_string(),
            amount,
            StockKind::Pool,
            0.0,
            false,
            BTreeMap::new(),
        )
        .unwrap()
    }

    #[test]
    fn empty_composition_defaults_to_1to1() {
        let s = pool("a", 1.0);
        assert_eq!(s.composition, BTreeMap::from([(Quantity::Carbon, 1.0)]));
    }

    #[test]
    fn rejects_non_finite_amount() {
        let e = Stock::new(
            "a".to_string(),
            "d".to_string(),
            Quantity::Carbon,
            "mol".to_string(),
            f64::NAN,
            StockKind::Pool,
            0.0,
            false,
            BTreeMap::new(),
        );
        assert!(matches!(e, Err(SimError::Validation(_))));
    }

    #[test]
    fn rejects_unclamped_pool() {
        let e = Stock::new(
            "a".to_string(),
            "d".to_string(),
            Quantity::Carbon,
            "mol".to_string(),
            1.0,
            StockKind::Pool,
            0.0,
            true,
            BTreeMap::new(),
        );
        assert!(matches!(e, Err(SimError::Validation(_))));
    }

    #[test]
    fn rejects_multi_quantity_population() {
        let e = Stock::new(
            "p".to_string(),
            "d".to_string(),
            Quantity::Carbon,
            "mol".to_string(),
            1.0,
            StockKind::Population,
            0.0,
            false,
            BTreeMap::from([(Quantity::Carbon, 1.0), (Quantity::Oxygen, 2.0)]),
        );
        assert!(matches!(e, Err(SimError::Validation(_))));
    }

    #[test]
    fn state_rejects_key_id_mismatch() {
        let s = pool("a", 1.0);
        let stocks = BTreeMap::from([("wrong".to_string(), s)]);
        let e = State::new(0, stocks, 0, BTreeMap::new());
        assert!(matches!(e, Err(SimError::Validation(_))));
    }
}
